### PR TITLE
Clusters checker

### PR DESCRIPTION
**Pull request by miiu96**
- Implemented a tool that will compare two elasticsearch clusters and log the difference in  the console 
- This tool has more console arguments to specify how to run the checks

### CODE REVIEW

File: tools/clusters-checker/cmd/checker/main.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
+package main
+
+import (
+	"github.com/ElrondNetwork/elastic-indexer-go/tools/clusters-checker/pkg/checkers"
```
Comment (bogdan-rosianu) [inline tools/clusters-checker/cmd/checker/main.go:4]: sort imports

File: tools/clusters-checker/cmd/checker/main.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+const configFileName = "config.toml"
+
+var (
+	log = logger.GetOrCreate("main")
+
+	// defines the path to the config folder
+	configPath = cli.StringFlag{
+		Name:  "config-path",
+		Usage: "The path to the config folder",
+		Value: "./",
```
Comment (bogdan-rosianu) [inline tools/clusters-checker/cmd/checker/main.go:24]: shouldn't the default value be `./config.toml`?

File: tools/clusters-checker/cmd/checker/main.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+		os.Exit(1)
+	}
+
+}
+
+func checkClusters(ctx *cli.Context) {
+	cfgPath := ctx.String(configPath.Name)
+	cfg, err := loadConfigFile(cfgPath)
+	if err != nil {
+		log.Error("cannot load config file", "error", err.Error())
```
Comment (bogdan-rosianu) [inline tools/clusters-checker/cmd/checker/main.go:76]: return here

File: tools/clusters-checker/cmd/checker/main.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	}
+
+	clusterChecker, err := checkers.CreateClusterChecker(cfg)
+	if err != nil {
+		log.Error("cannot create cluster checker", "error", err.Error())
```
Comment (bogdan-rosianu) [inline tools/clusters-checker/cmd/checker/main.go:81]: return here

File: tools/clusters-checker/pkg/checkers/count.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+	countDestinationCluster, err := cc.clientDestination.DoCountRequest(index, nil)
+	if err != nil {
+		return err
+	}
+
+	difference := int64(countSourceCluster) - int64(countDestinationCluster)
+
+	if difference == 0 {
+		log.Info("number of documents are the same", "index", index,
```
Comment (bogdan-rosianu) [inline tools/clusters-checker/pkg/checkers/count.go:30]: number of documents **is** the same

File: tools/clusters-checker/pkg/checkers/factory.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package checkers
+
+import (
+	"fmt"
```
Comment (bogdan-rosianu) [inline tools/clusters-checker/pkg/checkers/factory.go:4]: sort imports, there should be a space after "fmt"

File: tools/clusters-checker/pkg/checkers/factory.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/ElrondNetwork/elastic-indexer-go/tools/clusters-checker/pkg/client"
+	"github.com/ElrondNetwork/elastic-indexer-go/tools/clusters-checker/pkg/config"
+	"github.com/elastic/go-elasticsearch/v7"
+)
+
+func CreateClusterChecker(cfg *config.Config) (*clusterChecker, error) {
```
Comment (bogdan-rosianu) [inline tools/clusters-checker/pkg/checkers/factory.go:10]: add comments to exported functions

File: tools/clusters-checker/pkg/checkers/process.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (cc *clusterChecker) compareIndex(index string) error {
+	return nil
```
Comment (bogdan-rosianu) [inline tools/clusters-checker/pkg/checkers/process.go:29]: finish the implementation

File: tools/clusters-checker/pkg/client/count.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+// DoCountRequest will get the number of elements that correspond with the provided query
+func (esc *esClient) DoCountRequest(index string, body []byte) (uint64, error) {
+	res, err := esc.client.Count(
+		esc.client.Count.WithIndex(index),
+		esc.client.Count.WithBody(bytes.NewBuffer(body)),
+	)
+	if err != nil {
+		return 0, err
+	}
+	if err != nil {
```
Comment (bogdan-rosianu) [inline tools/clusters-checker/pkg/client/count.go:17]: duplicated if

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: tools/clusters-checker/cmd/checker/main.go (hunk 1 above)
Comment (miiu96) [reply tools/clusters-checker/cmd/checker/main.go]: done.

File: tools/clusters-checker/cmd/checker/main.go (hunk 2 above)
Comment (miiu96) [reply tools/clusters-checker/cmd/checker/main.go:25]: no. this is used to provide the path to the `config.toml` file

File: tools/clusters-checker/cmd/checker/main.go (hunk 4 above)
Comment (miiu96) [reply tools/clusters-checker/cmd/checker/main.go:135]: done.

File: tools/clusters-checker/pkg/checkers/factory.go (hunk 6 above)
Comment (miiu96) [reply tools/clusters-checker/pkg/checkers/factory.go:5]: done.

File: tools/clusters-checker/pkg/checkers/factory.go (hunk 7 above)
Comment (miiu96) [reply tools/clusters-checker/pkg/checkers/factory.go]: added

File: tools/clusters-checker/pkg/checkers/process.go (hunk 8 above)
Comment (miiu96) [reply tools/clusters-checker/pkg/checkers/process.go]: done.

File: tools/clusters-checker/pkg/client/count.go (hunk 9 above)
Comment (miiu96) [reply tools/clusters-checker/pkg/client/count.go:18]: removed.

File: tools/clusters-checker/pkg/checkers/process_indices_with_timestamp.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+	missingFromDestination map[string]json.RawMessage
+
+	clientSource         ESClient
+	clientDestination    ESClient
+	indicesNoTimestamp   []string
+	indicesWithTimestamp []string
+
+	startTimestamp, stopTimestamp int
+
+	logPrefix string
```
Comment (dragos-rebegea) [inline tools/clusters-checker/pkg/checkers/process_indices_with_timestamp.go:28]: You can use LoggerWithIdentifier for the same result but a cleaner implementation.

File: tools/clusters-checker/pkg/client/count.go (hunk 9 above)
Comment (dragos-rebegea) [reply tools/clusters-checker/pkg/client/count.go:18]: duplicated if still here

File: tools/clusters-checker/pkg/client/client.go [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+	"github.com/tidwall/gjson"
+)
+
+var (
+	log = logger.GetOrCreate("clusters-checker/pkg/client")
+
+	httpStatusesForRetry = []int{429, 502, 503, 504}
+)
+
+type esClient struct {
```
Comment (dragos-rebegea) [inline tools/clusters-checker/pkg/client/client.go:27]: I think its a good idea to extract the client into a separate tool that can be easier reused.

File: tools/clusters-checker/pkg/client/count.go (hunk 9 above)
Comment (miiu96) [reply tools/clusters-checker/pkg/client/count.go:18]: removed

File: tools/clusters-checker/pkg/checkers/process_indices_with_timestamp.go (hunk 10 above)
Comment (miiu96) [reply tools/clusters-checker/pkg/checkers/process_indices_with_timestamp.go:28]: In order to can use `LoggerWithIdentifier`  I have to include github.com/ElrondNetwork/elrond-eth-bridge
Will let as it is for the moment

**Review by dragos-rebegea — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**